File: CoreTests/color.cpp
#include "pch.hpp"

TEST(Color, ConversionRgbaHsva)
{
    {
        constexpr ColorRgba rgba = ColorRgba::Red();
        constexpr ColorHsva hsva = ColorHsva::Red();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Green();
        constexpr ColorHsva hsva = ColorHsva::Green();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Blue();
        constexpr ColorHsva hsva = ColorHsva::Blue();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Yellow();
        constexpr ColorHsva hsva = ColorHsva::Yellow();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Magenta();
        constexpr ColorHsva hsva = ColorHsva::Magenta();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::LightBlue();
        constexpr ColorHsva hsva = ColorHsva::LightBlue();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Black();
        constexpr ColorHsva hsva = ColorHsva::Black();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::White();
        constexpr ColorHsva hsva = ColorHsva::White();

        EXPECT_EQ(static_cast<ColorHsva>(rgba), hsva);
        EXPECT_EQ(static_cast<ColorRgba>(hsva), rgba);
    }
}

TEST(Color, ConversionRgbaf)
{
    {
        constexpr ColorRgba rgba = ColorRgba::Red();
        constexpr Colorf f = Colorf::Red();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Green();
        constexpr Colorf f = Colorf::Green();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Blue();
        constexpr Colorf f = Colorf::Blue();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Yellow();
        constexpr Colorf f = Colorf::Yellow();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Magenta();
        constexpr Colorf f = Colorf::Magenta();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::LightBlue();
        constexpr Colorf f = Colorf::LightBlue();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Black();
        constexpr Colorf f = Colorf::Black();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::Gray();
        constexpr Colorf f = Colorf::Gray();

        // This cannot be true as gray is defined with 0.5f with floats and 127 with bytes, so a rounding error occurs here
        EXPECT_NE(static_cast<Colorf>(rgba), f);
        EXPECT_NE(static_cast<ColorRgba>(f), rgba);
    }
    {
        constexpr ColorRgba rgba = ColorRgba::White();
        constexpr Colorf f = Colorf::White();

        EXPECT_EQ(static_cast<Colorf>(rgba), f);
        EXPECT_EQ(static_cast<ColorRgba>(f), rgba);
    }
}

TEST(Color, EqualityRgb)
{
    EXPECT_EQ(ColorRgb::LightBlue(), ColorRgb::LightBlue());
}

TEST(Color, EqualityRgba)
{
    EXPECT_EQ(ColorRgba::LightBlue(), ColorRgba::LightBlue());
}

TEST(Color, Equalityf)
{
    EXPECT_EQ(Colorf::LightBlue(), Colorf::LightBlue());
}

TEST(Color, EqualityHsva)
{
    EXPECT_EQ(ColorHsva::LightBlue(), ColorHsva::LightBlue());
}

File: CoreCSharpTests/src/Color.cs
using System;
using System.Diagnostics;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Xnor.Core;

namespace CoreCSharpTests
{
    [TestClass]
    public class Color
    {
        [TestClass]
        public class ConversionRgbaHsva
        {
            [TestMethod]
            public void Red()
            {
                ColorRgba rgba = ColorRgba.Red;
                ColorHsva hsva = ColorHsva.Red;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Green()
            {
                ColorRgba rgba = ColorRgba.Green;
                ColorHsva hsva = ColorHsva.Green;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Blue()
            {
                ColorRgba rgba = ColorRgba.Blue;
                ColorHsva hsva = ColorHsva.Blue;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Yellow()
            {
                ColorRgba rgba = ColorRgba.Yellow;
                ColorHsva hsva = ColorHsva.Yellow;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Magenta()
            {
                ColorRgba rgba = ColorRgba.Magenta;
                ColorHsva hsva = ColorHsva.Magenta;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void LightBlue()
            {
                ColorRgba rgba = ColorRgba.LightBlue;
                ColorHsva hsva = ColorHsva.LightBlue;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Black()
            {
                ColorRgba rgba = ColorRgba.Black;
                ColorHsva hsva = ColorHsva.Black;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void White()
            {
                ColorRgba rgba = ColorRgba.White;
                ColorHsva hsva = ColorHsva.White;
            
                Assert.AreEqual(hsva, (ColorHsva) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
        }
        
        [TestClass]
        public class ConversionRgbaf
        {
            [TestMethod]
            public void Red()
            {
                ColorRgba rgba = ColorRgba.Red;
                Colorf hsva = Colorf.Red;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Green()
            {
                ColorRgba rgba = ColorRgba.Green;
                Colorf hsva = Colorf.Green;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Blue()
            {
                ColorRgba rgba = ColorRgba.Blue;
                Colorf hsva = Colorf.Blue;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Yellow()
            {
                ColorRgba rgba = ColorRgba.Yellow;
                Colorf hsva = Colorf.Yellow;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Magenta()
            {
                ColorRgba rgba = ColorRgba.Magenta;
                Colorf hsva = Colorf.Magenta;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void LightBlue()
            {
                ColorRgba rgba = ColorRgba.LightBlue;
                Colorf hsva = Colorf.LightBlue;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void Black()
            {
                ColorRgba rgba = ColorRgba.Black;
                Colorf hsva = Colorf.Black;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
            
            [TestMethod]
            public void White()
            {
                ColorRgba rgba = ColorRgba.White;
                Colorf hsva = Colorf.White;
            
                Assert.AreEqual(hsva, (Colorf) rgba);
                Assert.AreEqual(rgba, (ColorRgba) hsva);
            }
        }
        
        [TestMethod]
        public void EqualityRgb()
        {
            Assert.AreEqual(ColorRgb.LightBlue, ColorRgb.LightBlue);
        }
        
        [TestMethod]
        public void EqualityRgba()
        {
            Assert.AreEqual(ColorRgba.LightBlue, ColorRgba.LightBlue);
        }
        
        [TestMethod]
        public void Equalityf()
        {
            Assert.AreEqual(Colorf.LightBlue, Colorf.LightBlue);
        }
        
        [TestMethod]
        public void EqualityHsva()
        {
            Assert.AreEqual(ColorHsva.LightBlue, ColorHsva.LightBlue);
        }
    }
}
